cns/restserver/ipam_test.go

```diff
@@ -1893,3 +1896,261 @@ func TestIPAMGetK8sInfinibandFailure(t *testing.T) {
 		t.Fatalf("Expected failing requesting IPs due to not able to set routes")
 	}
 }
+
+func TestIPAMGetStandaloneSWIFTv2(t *testing.T) {
+	svc := getTestService(cns.ServiceFabric)
+	middleware := middlewares.StandaloneSWIFTv2Middleware{}
+	svc.AttachIPConfigsHandlerMiddleware(&middleware)
+
+	orchestratorContext, _ := testPod1Info.OrchestratorContext()
+	mockMACAddress := "00:00:00:00:00:00"
+	mockGatewayIP := "10.0.0.1" // from mock wireserver gateway calculation on host subnet
+
+	tt := []struct {
+		name             string
+		req              cns.IPConfigsRequest
+		mockNMAgent      *fakes.NMAgentClientFake
+		expectedResponse *cns.IPConfigsResponse
+	}{
+		{
+			name: "Successful single IPAM for Standalone SwiftV2 pod, when NMAgent returns error for GetNCVersionList",
+			req: cns.IPConfigsRequest{
+				DesiredIPAddresses:  []string{testIP1},
+				OrchestratorContext: orchestratorContext,
+				PodInterfaceID:      testPod1Info.InterfaceID(),
+				InfraContainerID:    testPod1Info.InfraContainerID(),
+			},
+			mockNMAgent: &fakes.NMAgentClientFake{
+				GetNCVersionListF: func(_ context.Context) (nma.NCVersionList, error) {
+					// NMAgent returns an error, eg. NC is not programmed
+					return nma.NCVersionList{
+						Containers: []nma.NCVersion{},
+					}, errors.New("any NMAgent error")
+				},
+			},
+			expectedResponse: &cns.IPConfigsResponse{
+				Response: cns.Response{
+					ReturnCode: types.Success,
```

Comment:
i added a comment in the code's call site and duplicated the comment where we created the NMAgent mock in this test suite: https://github.com/Azure/azure-container-networking/pull/2932/files#diff-ce23fd49ec1c52dce46d6033e9c20161610d0ae18b005c76376fda090b5619ceR136-R138

essentially, for SwiftV2, it programs NICs and not NCs, so the current assumption, though "it works," is based on the wrong assumption. we need to confirm NIC programming with NMAgent and not NC programming. though we reuse NC concept within DNC and CNS, NMAgent doesn't care about it (ref. to one of the benefits of SwiftV2 design). for now we'll keep this behaviour and fix it in the future, ie. when NMAgent exposes a new API for NIC programming status. 

as it is, the risk is that a container may be delayed when coming up because SwiftV2 NIC goal state may not be in sync with host state, ie. NIC programming status on NMAgent

Reply:
So we make a call into NMAgent and it returns a failure, but we somehow expect this and recognize it as the expected response for the SWIFTv2 scenario?

Reply:
CNS will be configured for Standalone SwiftV2 and the NMAgent call is a side effect of getting NCs from CNS' internal state, so for now, it is accepted rather than expected. in the end, the response still has the NIC type and MAC address from the internal state and is returned successfully